sql/connect/server/src/main/scala/org/apache/spark/sql/connect/ml/MLUtils.scala

```diff
@@ -448,98 +454,149 @@ private[ml] object MLUtils {
-    "scale", // LinearRegressionModel
-    "coefficients",
-    "intercept",
-    "coefficientMatrix",
-    "interceptVector", // LogisticRegressionModel
-    "summary",
-    "hasSummary",
-    "evaluate", // LogisticRegressionModel
-    "evaluateEachIteration", // GBTClassificationModel
-    "predictions",
-    "predictionCol",
-    "labelCol",
-    "weightCol",
-    "labels", // _ClassificationSummary
-    "truePositiveRateByLabel",
-    "falsePositiveRateByLabel", // _ClassificationSummary
-    "precisionByLabel",
-    "recallByLabel",
-    "fMeasureByLabel",
-    "accuracy", // _ClassificationSummary
-    "weightedTruePositiveRate",
-    "weightedFalsePositiveRate", // _ClassificationSummary
-    "weightedRecall",
-    "weightedPrecision",
-    "weightedFMeasure", // _ClassificationSummary
-    "scoreCol",
-    "roc",
-    "areaUnderROC",
-    "pr",
-    "fMeasureByThreshold", // _BinaryClassificationSummary
-    "precisionByThreshold",
-    "recallByThreshold", // _BinaryClassificationSummary
-    "probabilityCol",
-    "featuresCol", // LogisticRegressionSummary
-    "objectiveHistory",
-    "coefficientStandardErrors", // _TrainingSummary
-    "degreesOfFreedom", // LinearRegressionSummary
-    "devianceResiduals", // LinearRegressionSummary
-    "explainedVariance", // LinearRegressionSummary
-    "meanAbsoluteError", // LinearRegressionSummary
-    "meanSquaredError", // LinearRegressionSummary
-    "numInstances", // LinearRegressionSummary
-    "pValues", // LinearRegressionSummary
-    "r2", // LinearRegressionSummary
-    "r2adj", // LinearRegressionSummary
-    "residuals", // LinearRegressionSummary
-    "rootMeanSquaredError", // LinearRegressionSummary
-    "tValues", // LinearRegressionSummary
-    "totalIterations", // LinearRegressionSummary
-    "k", // KMeansSummary
-    "numIter", // KMeansSummary
-    "clusterSizes", // KMeansSummary
-    "trainingCost", // KMeansSummary
-    "cluster", // KMeansSummary
-    "computeCost", // BisectingKMeansModel
-    "rank", // ALSModel
-    "itemFactors", // ALSModel
-    "userFactors", // ALSModel
-    "recommendForAllUsers", // ALSModel
-    "recommendForAllItems", // ALSModel
-    "recommendForUserSubset", // ALSModel
-    "recommendForItemSubset", // ALSModel
-    "associationRules", // FPGrowthModel
-    "freqItemsets" // FPGrowthModel
-  )
+  private lazy val ALLOWED_ATTRIBUTES = Seq(
+    (classOf[Identifiable], Array("toString")),
+
+    // Model Traits
+    (classOf[PredictionModel[_, _]], Array("predict", "numFeatures")),
+    (classOf[ClassificationModel[_, _]], Array("predictRaw", "numClasses")),
+    (classOf[ProbabilisticClassificationModel[_, _]], Array("predictProbability")),
+
+    // Summary Traits
+    (classOf[HasTrainingSummary[_]], Array("hasSummary", "summary")),
+    (classOf[TrainingSummary], Array("objectiveHistory", "totalIterations")),
+    (
+      classOf[ClassificationSummary],
+      Array(
+        "predictions",
```

Comment:
there will be so many duplicated "attribute"s given the class inheriting.

Reply:
yes